modules/nextflow/src/main/groovy/nextflow/cli/Launcher.groovy

```diff
@@ -593,17 +597,30 @@ class Launcher {
      */
     @PackageScope
     static void setProxy(String qualifier, Map<String,String> env ) {
-        assert qualifier in ['http','https','HTTP','HTTPS']
+        assert qualifier in ['http','https','ftp','HTTP','HTTPS','FTP']
         def str = null
         def var = "${qualifier}_" + (qualifier.isLowerCase() ? 'proxy' : 'PROXY')
 
         // -- setup HTTP proxy
         try {
-            List<String> proxy = parseProxy(str = env.get(var.toString()))
+            Map<String,String> proxy = parseProxy(str = env.get(var.toString()))
             if( proxy ) {
+                // avoid logging usernames and passwords
+                def username = proxy.remove('username')
+                def password = proxy.remove('password')
+
                 log.debug "Setting $qualifier proxy: $proxy"
-                System.setProperty("${qualifier.toLowerCase()}.proxyHost", proxy[0])
-                if( proxy[1] ) System.setProperty("${qualifier.toLowerCase()}.proxyPort", proxy[1])
+                System.setProperty("${qualifier.toLowerCase()}.proxyHost", proxy.host)
+                if( proxy.port ) System.setProperty("${qualifier.toLowerCase()}.proxyPort", proxy.port)
+                if( username && password ) {
+                    log.debug "Setting $qualifier proxy authenticator ..."
+                    Authenticator authenticator = new Authenticator() {
+                        protected PasswordAuthentication getPasswordAuthentication() {
+                            return new PasswordAuthentication( username, password.toCharArray() )
+                        }
+                    }
+                    Authenticator.setDefault(authenticator);
+                }
```

Comment:
What's the reason of this?

Reply:
It sets the authenticator (assuming a username and password is provided) that will be used to establish a network connection when the proxy server asks for authentication.

Reply:
My concern is how is determined to which host apply at these credentials? should not only be for the proxy? 

Also, why the previous username and password handling was not good?

Reply:
My use case is for Nextflow to be able to stage foreign files over HTTP/FTP from behind our institute's proxy server. These credentials are required by our proxy server only. Just so we're on the same page, would you be able to link me the previous username and password handling code? Maybe I missed it. Thanks!

Reply:
You may be [right](https://stackoverflow.com/a/1626616/395921), however, I fear setting the default authenticator can interfere with any HTTP connection. It May be needed to check the request type as shown in this comment https://stackoverflow.com/a/16340273/395921